### ISSUE TITLE
Likely host endianness bug in PFPO instruction

### ISSUE THREAD

**Issue by wrljet** [SDL-Hercules-390/hyperion#523]
Building `develop` branch on an s390x Linux system, the PFPO test fails.

```
>>>>> line 19321: Received unexpected wait state:  0002000180000000 000000000000DEAD
>>>>> line 19337: Storage compare mismatch.
                  Want: R:0000000000000710 416487ED 5110B461  
                  Got:  R:0000000000000710 00000000 00000000
>>>>> line 19339: Unexpected pgm type 0006.
Test "PFPO    ORIGINAL":  0 OK compares.  3 failures.
>>>>> line 19390: Received unexpected wait state:  0002000180000000 000000000000DEAD
>>>>> line 19406: Storage compare mismatch.
                  Want: R:0000000000000710 416487ED 5110B461  
                  Got:  R:0000000000000710 00000000 00000000
>>>>> line 19408: Unexpected pgm type 0006.
```

Building `develop` branch on an sparc64 NetBSD system, the PFPO test fails.

```
>>>>> line 19650: Received unexpected wait state:  0002000180000000 000000000000DEAD
>>>>> line 19666: Storage compare mismatch.
                  Want: R:0000000000000710 416487ED 5110B461
                  Got:  R:0000000000000710 00000000 00000000
>>>>> line 19668: Unexpected pgm type 0006.
Test "PFPO    ORIGINAL":  0 OK compares.  3 failures.
>>>>> line 19718: Received unexpected wait state:  0002000180000000 000000000000DEAD
>>>>> line 19735: Storage compare mismatch.
                  Want: R:0000000000000710 416487ED 5110B461
                  Got:  R:0000000000000710 00000000 00000000
>>>>> line 19737: Unexpected pgm type 0006.
```

**Comment by Fish-Git** [comment]
More than likely a host endianness bug. I suspect Bob's code is presuming a little-endian host. Let me look into it.

**Comment by Fish-Git** [comment]
Bill (@wrljet),

I don't have a big-endian test system I can test the following patch on, but FWIW, it works fine for Windows x64 (which of course is little-endian).

I would appreciate you giving it a quick test for me on your s390x Linux and sparc64 NetBSD systems for me before I commit it. Thanks.

* [pfpo.zip](https://github.com/SDL-Hercules-390/hyperion/files/10108081/pfpo.zip)

(p.s. you may or may not need to do a d2u (dos2unix) on the patch before applying it)

**Comment by wrljet** [comment]
Roger Wilco.

I'll get back to you.

**Comment by Fish-Git** [comment]
> Roger Wilco.
> 
> I'll get back to you.

Thanks.

Just as an FYI _(in case you or anyone else is interested)_, the actual fix is in the actual instruction function itself (`DEF_INST( perform_floating_point_operation )` starting near line 2215 after the patch is applied).

The other changes you see is just some additional minor cleanup I did as well.

Bob has a habit of using hard coded constants which makes his code much harder to understand and follow. There's much more I _could_ do to clean it up even more _(such as add some freaking **COMMENTS! (*)**)_, but  to be honest I'm just not in the mood. &nbsp; `:(`

___
**`(*)`** _It continues to amaze me how so many seemingly brilliant programmers fail to adequately comment their code! I've seen Linux kernel code and it's the same way! Sad._

**Comment by wrljet** [comment]
Goddamn, Fish, you're good!

1000 lines of diff!

PFPO tests now pass on s390x.

Sparc to follow...

**Comment by Fish-Git** [comment]
> 1000 lines of diff!

Like I said, it's _mostly_ cleanup. I changed a bunch of hard coded constants he was using to use the \#define constants defined in `esa390.h` instead.

The actual fix to the endian problem was fairly trivial, and is mostly just lines 2227-2232:

```C
    bool test_mode = regs->GR_L(0) & 0x80000000 ? true : false;
    BYTE otc       = regs->GR_LHHCH(0) & 0x7F;
    BYTE ofc1      = regs->GR_LHH(0)   & 0xFF;
    BYTE ofc2      = regs->GR_LHLCH(0);
    BYTE ccbits    = regs->GR_LHLCL(0);
    BYTE rm        = GR0_RM_BITS( ccbits );
```

THAT part was easy!

**Comment by wrljet** [comment]
Take the compliment for the odd time we're not cussing you.  :-)

**Comment by Fish-Git** [comment]
> Take the compliment for the odd time we're not cussing you.

I have my good days and bad days just like everyone else. &nbsp;`:)`

Sometimes they're mostly bad days, other times they're good days. Today just happened to be one of the good days. &nbsp;`:)

_**FYI:**_ &nbsp;There's still a rather _SERIOUS BUG_ that needs to be fixed that I just discovered: he's never setting the Condition Code! _(Oops!)_ &nbsp;`:(`

I'm looking into it...

**Comment by wrljet** [comment]
Fish,

PFPO tests pass, now as well, on Sparc64 CPU.

Excellent work!

Thanks,
Bill

**Comment by Fish-Git** [comment]
Fix committed (1af7890235b0a425e97dc3733d897ae8fa0b3b83). Closing.
